bindings/matrix-sdk-ffi/src/room_list.rs

```diff
@@ -192,62 +161,137 @@ impl RoomList {
+            unsafe { addr_of_mut!((*ptr).this).as_ref() }
+                // SAFETY: `this` contains a non null value.
+                .unwrap();
+
+        // Now we can create `entries_stream` and `dynamic_entries_controller` by
+        // borrowing `this`, which is going to live long enough since it will live as
+        // long as `entries_stream` and `dynamic_entries_controller`.
         let (entries_stream, dynamic_entries_controller) =
-            self.inner.entries_with_dynamic_adapters(
+            this.inner.entries_with_dynamic_adapters(
                 page_size.try_into().unwrap(),
-                self.room_list_service.inner.client().roominfo_update_receiver(),
+                client.roominfo_update_receiver(),
             );
 
-        RoomListEntriesWithDynamicAdaptersResult {
-            controller: Arc::new(RoomListDynamicEntriesController::new(
-                dynamic_entries_controller,
-                self.room_list_service.inner.client(),
-            )),
-            entries_stream: Arc::new(TaskHandle::new(RUNTIME.spawn(async move {
-                pin_mut!(entries_stream);
+        // FFI dance to make those values consumable by foreign language, nothing fancy
+        // here, that's the real code for this method.
+        let dynamic_entries_controller =
+            Arc::new(RoomListDynamicEntriesController::new(dynamic_entries_controller));
 
-                while let Some(diff) = entries_stream.next().await {
-                    listener.on_update(diff.into_iter().map(Into::into).collect());
-                }
-            }))),
+        let entries_stream = Arc::new(TaskHandle::new(RUNTIME.spawn(async move {
+            pin_mut!(entries_stream);
+
+            while let Some(diff) = entries_stream.next().await {
+                listener.on_update(diff.into_iter().map(Into::into).collect());
+            }
+        })));
+
+        // Initialize the second field `controller`.
+        //
+        // SAFETY: `ptr` is correctly aligned.
+        unsafe {
+            addr_of_mut!((*ptr).controller).write(dynamic_entries_controller);
         }
+
+        // Initialize the third and last field `entries_stream`.
+        //
+        // SAFETY: `ptr` is correctly aligned.
+        unsafe {
+            addr_of_mut!((*ptr).entries_stream).write(entries_stream);
+        }
+
+        // The result is complete, let's return it!
+        //
+        // SAFETY: `result` is fully initialized, all its fields have received a valid
+        // value.
+        Arc::new(unsafe { result.assume_init() })
     }
 
     fn room(&self, room_id: String) -> Result<Arc<RoomListItem>, RoomListError> {
         self.room_list_service.room(room_id)
     }
 }
 
-#[derive(uniffi::Record)]
-pub struct RoomListEntriesResult {
-    pub entries: Vec<RoomListEntry>,
-    pub entries_stream: Arc<TaskHandle>,
+#[derive(uniffi::Object)]
+pub struct RoomListEntriesWithDynamicAdaptersResult {
+    this: Arc<RoomList>,
```

Comment:
Should this be `Pin`ned, considering the requirement for the reference to never be invalidated?

Reply:
This is not a self-referential structure. We just need `this` to live as long as `entries_stream`, but `entries_stream` doesn't reference `this` (“pointer speaking”).